dvc/repo/diff.py

```diff
@@ -0,0 +1,178 @@
+            }
+        else:
+            results[path] = {
+                "b_output": b_outs[path],
+                "new_file": True,
+                "deleted_file": False,
+                "is_dir": b_outs[path].is_dir_cache,
+            }
+    return results
+
+
+def _diff_dir(self, target, diff_dct):
+    engine = inflect.engine()
+    msg = ""
+    a_entries, b_entries = {}, {}
+    if not diff_dct["new_file"]:
+        msg += "-{} with md5 {}\n".format(
+            target, diff_dct["a_output"].checksum
+        )
+        a_entries = _extract_output(self, diff_dct["a_output"])
+    if not diff_dct["deleted_file"]:
+        msg += "+{} with md5 {}\n".format(
+            target, diff_dct["b_output"].checksum
+        )
+        b_entries = _extract_output(self, diff_dct["b_output"])
+    msg += "\n"
+    result = _get_tree_changes(self, a_entries, b_entries)
+    msg += f"{result['ident']} " + engine.plural("file", result["ident"])
+    msg += " not changed, "
+    msg += f"{result['changes']} " + engine.plural("file", result["changes"])
+    msg += " modified, "
+    msg += f"{result['new']} " + engine.plural("file", result["new"])
+    msg += " added, "
+    msg += f"{result['del']} " + engine.plural("file", result["del"])
+    msg += " deleted, "
+    if result["diff_size"] == "0 Bytes":
+        msg += "size wasn't changed"
+    else:
+        msg += "size was " + "increased" * (result["diff_sign"] != "-")
+        msg += (
+            "decreased" * (result["diff_sign"] == "-")
+            + " by "
+            + result["diff_size"]
+        )
+    return msg
+
+
+def _diff_file(self, target, diff_dct):
+    msg = ""
+    size = 0
+    if not diff_dct["new_file"]:
+        msg += "-{} with md5 {}\n".format(
+            target, diff_dct["a_output"].checksum
+        )
+        size -= os.path.getsize(
+            self.cache.local.get(diff_dct["a_output"].checksum)
+        )
+    if not diff_dct["deleted_file"]:
+        msg += "+{} with md5 {}\n\n".format(
+            target, diff_dct["b_output"].checksum
+        )
+        size += os.path.getsize(
+            self.cache.local.get(diff_dct["b_output"].checksum)
+        )
+    if not diff_dct["new_file"] and not diff_dct["deleted_file"] and size == 0:
+        msg += "file was not changed"
+    elif diff_dct["new_file"]:
+        msg += "added file with size {}".format(humanize.naturalsize(size))
+    elif diff_dct["deleted_file"]:
+        msg += "deleted file with size {}".format(humanize.naturalsize(size))
+    else:
+        msg += "file was modified, file size " + "increased" * (size >= 0)
+        msg += "decreased" * (size < 0) + " by " + humanize.naturalsize(size)
+    return msg
+
+
+def _diff_royal(self, target, diff_dct):
+    msg = "diff for " + target + "\n"
+    if diff_dct["is_dir"]:
+        return _diff_dir(self, target, diff_dct)
+    return msg + _diff_file(self, target, diff_dct)
+
+
+def diff(self, target, a_ref=None, b_ref=None):
```

Comment:
The biggest concern I have - the interface should not be string based. We should make it return some meaningful class instance that describes the difference and that can then print itself. Otherwise we are mixing logic that collects the diff with the logic that represents it. It complicates the code with all these appends and string manipulation, makes it hard to support, etc.

Reply:
Sure, one question here, what is the best place to locate that class definition?

Reply:
I used list of dictionaries for now.